Add operator for logging device data and metadata

glopesdev opened the pull request:
For compliance with the Harp logging standard (https://github.com/harp-tech/protocol/issues/41) we require the device metadata to be saved next to the data to allow building automatic data readers.

Here we introduce a per-device variant of the `MessageWriter` operator called `DeviceDataWriter` which saves the device metadata as a subscription side-effect and ensures all registers are demuxed and named correctly when stored into the specified target path.

glopesdev commented on interface/Device.tt:
```diff
@@ -125,6 +125,97 @@ foreach (var register in deviceRegisters)
+            return Observable.Create<TSource>(observer =>
+            {
+                File.WriteAllText(metadataPath, Device.Metadata);
+                return source.SubscribeSafe(observer);
```
Depends on when we want the side-effect to run. If we move it to the observable construction stage, then putting this inside a `CreateObservable` would immediately save the metadata, rather than wait until the actual subscription.

bruno-f-cruz commented on interface/Device.tt:
```diff
@@ -125,6 +125,97 @@ foreach (var register in deviceRegisters)
+        [Category(nameof(Device.Metadata))]
+        [FileNameFilter("YAML Files (*.yml;*.yaml)|*.yml;*.yaml|All Files|*.*")]
+        [Editor("Bonsai.Design.SaveFileNameEditor, Bonsai.Design", DesignTypes.UITypeEditor)]
+        public string MetadataFileName { get; set; } = "device.yml";
```
oh thats right, sorry got confused with the core writer... my bad!

glopesdev commented:
@bruno-f-cruz Just to keep track of the changes, I renamed the operator as `DeviceDataWriter` and refactored its behavior to simply enforce the standard on a per-folder level.
